File: pc/src/bme280_tool/records.py
from dataclasses import dataclass
from datetime import datetime, timedelta

FLAG_TIME_VALID = 0x01
TIME_FMT = "%Y-%m-%d %H:%M:%S"
CSV_HEADER = ["timestamp", "temp_c", "hum_pct", "pressure_pa", "time_valid", "boot_id"]


@dataclass(frozen=True)
class LogRecord:
    timestamp: int
    temp_centi: int
    hum_centi: int
    pressure_pa: int
    flags: int
    boot_id: int

    @classmethod
    def from_array(cls, a: list) -> "LogRecord":
        return cls(*a)

    @property
    def time_valid(self) -> bool:
        return bool(self.flags & FLAG_TIME_VALID)

    @property
    def temp_c(self) -> float:
        return self.temp_centi / 100

    @property
    def hum_pct(self) -> float:
        return self.hum_centi / 100

    @property
    def pressure_hpa(self) -> float:
        return self.pressure_pa / 100


@dataclass(frozen=True)
class SyncInfo:
    """Device clock state captured at connect time, used to place unsynced records on the PC timeline."""

    boot_id: int
    uptime_s: int
    pc_time: datetime


def resolve_time(rec: LogRecord, sync: SyncInfo) -> datetime | None:
    if rec.time_valid:
        return datetime.fromtimestamp(rec.timestamp)
    if rec.boot_id == sync.boot_id:
        return sync.pc_time - timedelta(seconds=sync.uptime_s - rec.timestamp)
    return None


def _relative_label(rec: LogRecord) -> str:
    h, rem = divmod(rec.timestamp, 3600)
    m, s = divmod(rem, 60)
    return f"boot#{rec.boot_id} +{h:02d}:{m:02d}:{s:02d}"


def format_time(rec: LogRecord, sync: SyncInfo) -> str:
    dt = resolve_time(rec, sync)
    if dt is None:
        return _relative_label(rec)
    prefix = "" if rec.time_valid else "~"
    return prefix + dt.strftime(TIME_FMT)


def csv_row(rec: LogRecord, sync: SyncInfo) -> list[str]:
    dt = resolve_time(rec, sync)
    when = dt.isoformat() if dt is not None else _relative_label(rec)
    return [when, f"{rec.temp_c:.2f}", f"{rec.hum_pct:.2f}", str(rec.pressure_pa), "1" if rec.time_valid else "0", str(rec.boot_id)]

File: pc/internal/records/records.go
// Package records decodes device log records and places them on the PC timeline.
package records

import (
	"fmt"
	"strconv"
	"time"
)

const (
	FlagTimeValid = 0x01
	TimeFormat    = "2006-01-02 15:04:05"
	csvTimeFormat = "2006-01-02T15:04:05"
)

var CSVHeader = []string{"timestamp", "temp_c", "hum_pct", "pressure_pa", "time_valid", "boot_id"}

type LogRecord struct {
	Timestamp  int64 // epoch seconds if TimeValid, else seconds since boot
	TempCenti  int
	HumCenti   int
	PressurePa int64
	Flags      int
	BootID     int
}

// FromArray decodes the protocol's [timestamp, temp_centi, hum_centi, pressure_pa, flags, boot_id].
func FromArray(a []int64) (LogRecord, error) {
	if len(a) != 6 {
		return LogRecord{}, fmt.Errorf("record has %d fields, want 6", len(a))
	}
	return LogRecord{Timestamp: a[0], TempCenti: int(a[1]), HumCenti: int(a[2]), PressurePa: a[3], Flags: int(a[4]), BootID: int(a[5])}, nil
}

func (r LogRecord) TimeValid() bool      { return r.Flags&FlagTimeValid != 0 }
func (r LogRecord) TempC() float64       { return float64(r.TempCenti) / 100 }
func (r LogRecord) HumPct() float64      { return float64(r.HumCenti) / 100 }
func (r LogRecord) PressureHPa() float64 { return float64(r.PressurePa) / 100 }

// SyncInfo is the device clock state captured at connect time.
type SyncInfo struct {
	BootID  int
	UptimeS int64
	PCTime  time.Time
}

// ResolveTime returns the record's wall-clock time when it is known (synced) or estimable
// (same boot as the current connection); ok is false for records from earlier boots.
func ResolveTime(r LogRecord, s SyncInfo) (time.Time, bool) {
	if r.TimeValid() {
		return time.Unix(r.Timestamp, 0), true
	}
	if r.BootID == s.BootID {
		return s.PCTime.Add(-time.Duration(s.UptimeS-r.Timestamp) * time.Second), true
	}
	return time.Time{}, false
}

func relativeLabel(r LogRecord) string {
	h := r.Timestamp / 3600
	m := (r.Timestamp % 3600) / 60
	sec := r.Timestamp % 60
	return fmt.Sprintf("boot#%d +%02d:%02d:%02d", r.BootID, h, m, sec)
}

func FormatTime(r LogRecord, s SyncInfo) string {
	t, ok := ResolveTime(r, s)
	if !ok {
		return relativeLabel(r)
	}
	if r.TimeValid() {
		return t.Format(TimeFormat)
	}
	return "~" + t.Format(TimeFormat)
}

func CSVRow(r LogRecord, s SyncInfo) []string {
	when := relativeLabel(r)
	if t, ok := ResolveTime(r, s); ok {
		when = t.Format(csvTimeFormat)
	}
	valid := "0"
	if r.TimeValid() {
		valid = "1"
	}
	return []string{when, fmt.Sprintf("%.2f", r.TempC()), fmt.Sprintf("%.2f", r.HumPct()),
		strconv.FormatInt(r.PressurePa, 10), valid, strconv.Itoa(r.BootID)}
}
